### DISCUSSION TITLE
Using a local module file

### DISCUSSION THREAD

**Discussion by MinusOneTwelfth-1** [fonsp/Pluto.jl#1728] [Questions about using Pluto] [answered]
I have made a simple module file, myModule.jl in the same folder as the Pluto notebook. This module **export**s a function called myFunction.

In the first cell I have:
```
begin
	include("myModule.jl")
	using .myModule  
	
end
```
And in the next cell:

`myFunction(3)`

After I Shift+Enter that, I get this error:
```
UndefVarError: myFunction not defined

    top-level scope@Local: 1
```
... But now if I go back to the top cell and just do Shift+Enter, the second cell shows the correct result.

### **To sum up,..**
 After entering myFunction into any cell (or after doing Shift+Enter in an existing cell ) I need to go back to the first cell and do a Shift+Enter again to refresh the module's exports.

I'm working around this by putting `myFunction = myModule.myFunction` in the first cell, but it would be a pain if I have lots more functions.

Is this a bug or is there something I can do to to make it work?

**Comment by icweaver** [comment, marked as answer]
Hey @MinusOneTwelfth-1! As far as I can tell, `include[t]`ing things in Pluto [can get a bit dicey](https://github.com/fonsp/Pluto.jl/issues/1101) sometimes. When I get to the stage of working with modules, I usually just turn what I am working on into a package, which makes a lot of really nice things automatically available, such as the [Revise.jl](https://timholy.github.io/Revise.jl/stable/#Usage-example-1) workflow. Using your example above, I would do something like the following:

## 1. Create a skeleton package

[PkgTemplates.jl](https://github.com/invenia/PkgTemplates.jl) is awesome for this, and super customizable:

```shell
pkg> add PkgTemplates
julia> using PkgTemplates
julia> Template(dir=".")("MyPkg") # This assumes you have your global git configured already
```

## 2. Add `Revise.jl` globally
### (or just to your new package, up to you)

```julia
pkg> add Revise 
```

## 3. Fire up a `Pluto.jl` notebook

It doesn't really matter where you create the notebook as long as you point it to your package, but I just like keeping everything self-contained in the top-level of my package, so it might look something like this:

```
MyPkg
├── LICENSE
├── Manifest.toml
├── notebook_MyPkg.jl
├── Project.toml
├── README.md
├── src
│   └── MyPkg.jl
└── test
    └── runtests.jl
``` 

Now over in my `Pluto.jl` notebook (`notebook_MyPkg.jl`), if I have a cell like this:

```julia
begin
    import Pkg
    Pkg.activate(Base.current_project())
    using Revise, MyPkg
end
```

any changes that I make to the package should be automatically reflected after re-running that cell. For example, if I edit `MyPkg/src/MyPkg.jl` from:

```julia
module MyPkg

export my_function

my_function(x) = x^2

end
```

to:

```julia
module MyPkg

export my_function

my_function(x) = x^3

end
```

and re-run a `Pluto.jl` cell with the following:

```julia
my_function(3)
```

the output should update accordingly

**Reply by MinusOneTwelfth-1** [reply]
Thanks, icweaver  ... I had been hesitating to dive into a package based workflow because it seemed a bit daunting as a newcomer to Julia & Pluto. But with your explanations I'm finding it easy to packagize my Pluto work and to use modules etc optimally.

Thanks again.

**Comment by floswald** [comment]
hey there, this does not seem to work for me anymore. I get this

```
begin
    import Pkg
    Pkg.activate(Base.current_project())
    using Revise, Plutotest2
end
```

```
MethodError: no method matching activate(::Nothing)

Closest candidates are:

activate(; temp, shared, io) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:1430

activate(!Matched::AbstractString; shared, temp, io) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:1456

activate(!Matched::Function, !Matched::AbstractString) at /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/Pkg/src/API.jl:1497

    top-level scope@Local: 3
```

it seems the `current_project()` does return `nothing` instead of the required path.

**Reply by icweaver** [reply]
Howdy! Could you verify where your notebook is being saved? I think `current_project` expects a `Project.toml` to be somewhere in the parent or current directory of your notebook. If it's a new notebook, then it might have been created in a temporary location, which would explain your error

**Reply by floswald** [reply]
no idea what happened but it works now. 🤷 did i run this from the vscode terminal before? no idea. anyway, fixed. :-)
